rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md

```diff
@@ -0,0 +1,112 @@
+# RFC 6330 - 2021-04-07 - Emitting multiple log events from remap transform
+
+This RFC describes an apporach for emitting multiple log events from a remap transform.
+
+## Scope
+
+In:
+
+- Emitting multiple log events from a remap transform
+
+Out:
+
+- Emitting multiple metrics from a remap transform
+
+## Motivation
+
+Currently the remap transform can only modify an event as it is passing through. Users have asked [1][] [2][] to be able to transform an incoming event into multiple events. This is useful, for example, when parsing an incoming JSON payload that contains an array for which you'd like to publish an event for each element.
+
+Currently, users are restricted to using the Lua or WASM transform for this, [introducing a substantial bottleneck](https://user-images.githubusercontent.com/316880/105531520-d3643200-5cbf-11eb-8b35-fe1c99e5c254.png).
+
+Additionally, I think we can leverage the approach I describe here in the future to allow users to emit different data types than the incoming type. For example, users could emit logs related to an incoming metric, or metrics from an incoming log.
+
+## Internal Proposal
+
+This is roughly the same as https://github.com/timberio/vector/issues/6330#issuecomment-772562955 with some slight tweaks.
+
+A new `emit_log` function will be added to the VRL stdlib:
+
+```text
+emit_log(value: Object)
+```
+
+This function will cause the object passed as value to be emitted at that point and flushed downstream. The emitted log will have its metadata copied from the input event.
+
+Additionally, an `emit_root` (we can work on the naming) config option will be added to the `remap` transform to configure whether `.` is emitted after the transform runs. It will default to `true` to preserve the current behavior but can be set to `false` by users to supress this behavior. Admittedly, I'm not wild about introducing this additional config option, but I'm not seeing another great alternative.
+
+This will be able to be combined with the iteration mechanism that will be introduced [#6031](https://github.com/timberio/vector/issues/6031) to emit an unknown number of events. Naively this might look something like:
+
+```text
+for stooge in .stooges
+   emit(stooge)
+end
+```
+
+In the future we can also add functions for emitting metrics like:
+
+```text
+emit_counter(namespace: String, name: String, timestamp: Timestamp, value: Float, kind: "absolute"|"relative")
+```
+
+I considered having just an `emit_metric()` but it would require users to pass in objects that match exactly the internal representation we have for metrics.
```

Comment:
Not necessarily. The Lua wrapper has at various points converted to/from the internal format as necessary, as does the WASM wrapper. There's no reason this couldn't either.

Reply:
Yeah, that's true, but we don't have a concept of structs in VRL so it would depend on the user passing a map of key/values that matches some schema.